benchmark_test/benchmark_test.go

```diff
@@ -0,0 +1,160 @@
+		ElasticsearchAuth:    os.Getenv("ELASTICSEARCH_AUTH"),
+		ApmElasticsearchUrl:  getEnv("APM_ELASTICSEARCH_URL", "http://localhost:9200"),
+		ApmElasticsearchAuth: os.Getenv("APM_ELASTICSEARCH_AUTH"),
+		RunTimeout:           conv.ToDuration(getEnv("RUN_TIMEOUT", "3m")),
+		FlushTimeout:         time.Second * 10,
+		TransactionFrequency: time.Millisecond,
+		TransactionLimit:     math.MaxInt32,
+		SpanMaxLimit:         10,
+		SpanMinLimit:         10,
+		ErrorFrequency:       time.Millisecond,
+		ErrorLimit:           0,
+		ErrorFrameMaxLimit:   10,
+		ErrorFrameMinLimit:   10,
+	}
+
+	var err error
+	glReportNode, err = es.NewConnection(glInput.ElasticsearchUrl, glInput.ElasticsearchAuth)
+	if err != nil {
+		panic(errors.Wrap(err, "Elasticsearch not reachable, won't be able to index a report"))
+	}
+	glWarmup = conv.ToDuration(getEnv("WARM_UP", "1m"))
+	glCooldown = conv.ToDuration(getEnv("COOL_DOWN", "1m"))
+	glRegressionMargin = conv.ToFloat64(getEnv("REGRESSION_MARGIN", "1.1"))
+	glRegressionDays = getEnv("REGRESSION_DAYS", "7")
+
+	fmt.Println(fmt.Sprintf("regression margin is %.1f", glRegressionMargin))
+	fmt.Println(fmt.Sprintf("regression days is %s", glRegressionDays))
+	// prevent send on closed chan error
+	apm.DefaultTracer.Close()
+
+	warmUp()
+	code := m.Run()
+	os.Exit(code)
+}
+
+func TestTransactionsOnly(t *testing.T) {
+	report := worker.Run(
+		builder().WithSpans(0).Input)
+	verify(t, report)
+}
+
+func TestTransactionsWithSpans(t *testing.T) {
+	report := worker.Run(builder().Input)
+	verify(t, report)
+}
+
+func TestTransactionsWithManySpans(t *testing.T) {
+	report := worker.Run(
+		builder().WithSpans(50).Input)
+	verify(t, report)
+}
+
+func TestErrors(t *testing.T) {
+	report := worker.Run(
+		builder().WithoutTransactions().WithErrors().Input)
+	verify(t, report)
+}
+
+func TestErrorsWithManyFrames(t *testing.T) {
+	report := worker.Run(
+		builder().WithoutTransactions().WithErrorFrames(100).Input)
+	verify(t, report)
+}
+
+func TestHighLoad(t *testing.T) {
+	report := worker.Run(
+		builder().WithErrorFreq(time.Microsecond).WithTransactionFreq(time.Microsecond).Input)
+	verify(t, report)
+}
+
+// asserts there are no performance regressions for a given workload
+func verify(t *testing.T, report models.Report) {
+	coolDown()
+	assert.True(t, report.EventsIndexed > 100, fmt.Sprintf("not enough events indexed: %d", report.EventsIndexed))
+	err := VerifyNoRegression(glReportNode, report, glRegressionMargin, glRegressionDays)
+	if !assert.NoError(t, err, err) {
+		t.Log("Report ==========")
+		t.Log(fmt.Sprintf("%+v", report))
+	}
+}
+
+// sends a moderate load to apm-server
+func warmUp() {
+	input := builder().WithErrors().Input
+	input.RunTimeout = glWarmup
+	input.ElasticsearchUrl = ""
+	fmt.Println(fmt.Sprintf("warming up %.1f seconds...", glWarmup.Seconds()))
+	worker.Run(input)
+	coolDown()
+}
+
+// waits an arbitrary time for events in elasticsearch be flushed
+func coolDown() {
```

Comment:
there is a `flushTimeout` name, and a `flush` function somewhere else, i wanted to avoid ambiguity where it was easy to avoid. it is not just waiting for ES flush, but apm-server to free some memory, etc.

Reply:
fair enough, wasn't aware of the cooling down of the server.